### PR TITLE
Introduce new universe v3 repo schema.

### PR DESCRIPTION
**Pull request by BenWhitehead**
The following changes were made to the project structure to ensure we're
able to handle multiple versions of universe schema in the same
codebase.

* Introduce new `cosmos-json` module that contains circe
  endcoders/decoders for models defined in cosmos-model
* Create new package `com.mesosphere.universe.v2.model` to contain all the
  existing universe model objects and schemas
* Create new package `com.mesosphere.universe.v3.model` to contain all the
  new universe v3 repo model objects and schemas
* Restructure the way subsections are represented in scala packages, this
  helps reduce the number of ambiguous import scopes created in code
  (since you can do a "partial import" in scala and then import from the
  scope that has already been imported) the following pattern change
  took place:
  * `com.mesosphere.cosmos.model.thirdparty.adminrouter` -> `com.mesosphere.cosmos.thirdparty.adminrouter.model`
  * `com.mesosphere.cosmos.model.thirdparty.marathon` -> `com.mesosphere.cosmos.thirdparty.marathon.model`
  * `com.mesosphere.cosmos.model.thirdparty.mesos.master` -> `com.mesosphere.cosmos.thirdparty.mesos.master.model`
* Introduce a new encoder/decoder for each subset of the model
  * `com.mesosphere.universe.v2.model`
  * `com.mesosphere.universe.v3.model`
  * `com.mesosphere.cosmos.thirdparty.adminrouter.model`
  * `com.mesosphere.cosmos.thirdparty.marathon.model`
  * `com.mesosphere.cosmos.thirdparty.mesos.master.model`

### CODE REVIEW

**Comment by BenWhitehead** [comment]
About 3,500 of the added lines is because of a new json file I added to the repo to test decoding of a v3 repo.

**Comment by BenWhitehead** [comment]
@cruhland please review

File: cosmos-json/src/main/scala/com/mesosphere/universe/v3/circe/Decoders.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  }
+  implicit val decodeV3PackagingVersion: Decoder[V3PackagingVersion] = {
+    Decoder.decodeString.map(V3PackagingVersion(_))
+  }
+
+  implicit val decodePackageDefinition: Decoder[PackageDefinition] = Decoder.instance[PackageDefinition] { (hc: HCursor) =>
+    hc.downField("packagingVersion").as[String].flatMap {
+      case "2.0" => hc.as[V2Package]
+      case "3.0" => hc.as[V3Package]
+    }
```
Comment (cruhland) [inline cosmos-json/src/main/scala/com/mesosphere/universe/v3/circe/Decoders.scala]: Could you add a test to make sure that we handle pattern match failure here in a "nice" way?

File: cosmos-model/src/main/scala/com/mesosphere/universe/v3/PackageDefinition.scala [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      s"Value '$value' does not conform to expected format $packageDetailsTagRegex"
+    )
+  }
+  object Tag {
+    val packageDetailsTagRegex = "^[^\\s]+$"
+    val packageDetailsTagPattern = Pattern.compile(packageDetailsTagRegex)
+  }
+
+  case class ReleaseVersion(value: Int) {
+    assert(value >= 0, s"Value $value is not >= 0")
```
Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/PackageDefinition.scala]: Do we have tests for decoding all of the above case classes that have `assert`s?

File: cosmos-json/src/test/scala/com/mesosphere/universe/v3/circe/RepositoryEncoderDecoderSpec.scala [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+          "cool-package",
+          Version("3.2.1"),
+          ReleaseVersion(2),
+          "bill@cool.co",
+          "some awesome package",
+          List(Tag("abc"), Tag("def")),
+          selected = Some(false),
+          marathon = Some(Marathon(ByteBuffer.wrap("testing".getBytes(StandardCharsets.UTF_8)))),
+          minDcosReleaseVersion = Some(DcosReleaseVersionParser.parseUnsafe("1.8"))
+        )
```
Comment (cruhland) [inline cosmos-json/src/test/scala/com/mesosphere/universe/v3/circe/RepositoryEncoderDecoderSpec.scala]: Could you update this test or add another test case that verifies we can encode/decode *all* the package fields? This only covers a subset.

**Comment by cruhland** [comment]
@BenWhitehead a few comments around testing, otherwise looks good.

File: cosmos-model/src/main/scala/com/mesosphere/universe/v3/PackageDefinition.scala (hunk 2 above)
Comment (BenWhitehead) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/PackageDefinition.scala]: New test added, preview of new error message:

File: cosmos-json/src/main/scala/com/mesosphere/universe/v3/circe/Decoders.scala (hunk 1 above)
Comment (BenWhitehead) [inline cosmos-json/src/main/scala/com/mesosphere/universe/v3/circe/Decoders.scala]: New test added, preview of new error message:
`Supported packagingVersion: [2.0, 3.0] but was: '3.1': El(DownField(packagingVersion),true),El(DownArray,true),El(DownField(packages),true)`

**Comment by cruhland** [comment]
:ship: it

**Pull request merged**